api4/channel.go

```diff
@@ -399,7 +399,7 @@ func getChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	if channel.Type == model.CHANNEL_OPEN {
-		if !c.App.SessionHasPermissionToTeam(c.App.Session, channel.TeamId, model.PERMISSION_READ_PUBLIC_CHANNEL) {
+		if !c.App.SessionHasPermissionToTeam(c.App.Session, channel.TeamId, model.PERMISSION_READ_PUBLIC_CHANNEL) && !c.App.SessionHasPermissionToChannel(c.App.Session, c.Params.ChannelId, model.PERMISSION_READ_CHANNEL) {
```

Comment:
I'm guessing this is needed in the context of guest accounts, such that a user might now have access to a team, but still won't have access to an open channel? Is the team check necessary at all now, given that check on the channel instead? Is the `channel.Type == model.CHANNEL_OPEN` test necessary at all given that both public and private channels require `model.PERMISSION_READ_CHANNEL`?

Reply:
Yes, I think so, one thing is the posibility to read public channels, which allows you to get the channel information, no matter if you are a member of the channel or not (you don't have that possibility for private channels). And other thing is the possibility of read the channels that you are a member of. Here we are allowing people to read public channels if they are members of the channel and have the permission to read the channel, or, if they are not member of the channel but have the permission to read public channels in the team. A regular user will have the permission to read public channels, and the permission to read specific channels where he/she is a member of the channel. The guest accounts only have the latter, so they will be able only to see the channels that they belongs to.

Reply:
This makes sense to me. However, without the explanation from Jesus, it's confusing. I think we either need to consider renaming the permissions in code (not sure about this because it'll break the assumption that the constant name and the permission name in the database are the same and probably lead to issues later) or clearly document what each permission means - in this case the names are not really self-documenting enough to understand what's going on.

Reply:
Note also: `PERMISSION_READ_PUBLIC_CHANNEL` is a Team-Scoped permission and `PERMISSION_READ_CHANNEL` is channel scoped. Could this be replaced by allowing `PERMISSION_READ_PUBLIC_CHANNEL` at the Channel level and just changing the permission checks to use `SessionHasPermissionToChannel` for it instead? Would this work? Would it be better? Would it be worse? Difficult!

Reply:
We can think about it, but I think it would be another user story. Probably we can remove the concept of `PERMISSION_READ_CHANNEL` and replace with the `PERMISSION_READ_PUBLIC_CHANNEL`, and `PERMISSION_READ_PRIVATE_CHANNEL`, at channel level, giving to `channel_user` the `PERMISSION_READ_PRIVATE_CHANNEL`, and to the team_user the `PERMISSION_READ_PUBLIC_CHANNEL`. It should work right away, I think. And the migration itself should be pretty simple (I think).

Reply:
Makes sense. I'm happy to defer this to another story - that will also give us more time to think about if there would also be any *unwanted* side effects of making that change.